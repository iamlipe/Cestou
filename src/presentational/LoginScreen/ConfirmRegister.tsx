import React, {useEffect} from 'react';
import styled from 'styled-components/native';
import {t} from 'i18next';
import {RouteProp, useNavigation, useRoute} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {AuthStackParamList} from '@/routes/stacks/AuthStack';

import Producer from '@/assets/svgs/welcome-producer.svg';
import Consumer from '@/assets/svgs/welcome-consumer.svg';

import Button from '@/components/Button';

type NavProps = NativeStackNavigationProp<AuthStackParamList, 'Onboarding'>;

type ParamList = {
  params: {
    userType: 'producer' | 'consumer';
    phoneOrEmail: string;
    password: string;
  };
};

export const ConfirmRegister = () => {
  const {navigate, addListener} = useNavigation<NavProps>();
  const route = useRoute<RouteProp<ParamList, 'params'>>();

  useEffect(() => {
    addListener('beforeRemove', event => {
      event.preventDefault();
    });
  });

  return (
    <StyledContainerScroll
      contentContainerStyle={{alignItems: 'center'}}
      showsVerticalScrollIndicator={false}>
      {route.params.userType === 'producer' ? (
        <StyledProducer testID="welcome-producer" />
      ) : (
        <StyledConsumer testID="welcome-consumer" />
      )}
      <StyledTitle>
        {t('text.screenConfirmRegister.registrationSuccessfully')}
      </StyledTitle>
      <StyledSubtitle>
        {t('text.screenConfirmRegister.validateYourRegistration')}
      </StyledSubtitle>
      <StyledConfirmButton
        testID="confirm-button"
        title={t('button.start')}
        onPress={() =>
          navigate('Onboarding', {
            phoneOrEmail: route.params.phoneOrEmail,
            password: route.params.password,
          })
        }
      />
    </StyledContainerScroll>
  );
};

const StyledContainerScroll = styled.ScrollView`
  min-height: 100%;
  background-color: ${({theme}) => theme.colors.BACKGROUND};
`;

const StyledProducer = styled(Producer)`
  margin-top: 40px;
`;

const StyledConsumer = styled(Consumer)`
  margin-top: 40px;
`;

const StyledTitle = styled.Text`
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.LARGER};
  color: ${({theme}) => theme.colors.PRIMARY_800};

  margin-top: 40px;
`;

const StyledSubtitle = styled.Text`
  width: 90%;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  text-align: center;
  color: ${({theme}) => theme.colors.GRAY_900};

  margin-top: 40px;
`;

const StyledConfirmButton = styled(Button)`
  margin-top: 40px;
`;
