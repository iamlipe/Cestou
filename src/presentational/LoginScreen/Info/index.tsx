import React, {memo} from 'react';
import styled from 'styled-components/native';
import {Linking} from 'react-native';
import {OUR_SITE} from 'react-native-dotenv';
import {useTranslation} from 'react-i18next';

import Button from '@/components/Button';

const Info = () => {
  const {t} = useTranslation();

  return (
    <StyledContainerInfo>
      <StyledText>{t('text.componentInfo.title')}</StyledText>
      <Button
        onPress={async () => Linking.openURL(OUR_SITE)}
        size="small"
        buttonColor="text_only"
        textColor="primary"
        title={t('button.goToSite')}
        noMargin
      />
    </StyledContainerInfo>
  );
};

const StyledContainerInfo = styled.View`
  width: 90%;
  align-self: center;
  justify-content: space-between;
  align-items: center;
  border: 1px solid ${({theme}) => theme.colors.PRIMARY_800};
  border-radius: 5px;
  padding: 8px;
  margin-bottom: 50px;
`;

const StyledText = styled.Text`
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  color: ${({theme}) => theme.colors.GRAY_900};
  margin-bottom: 16px;
`;

export default memo(Info);
