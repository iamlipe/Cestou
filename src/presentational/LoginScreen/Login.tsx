import React from 'react';
import styled from 'styled-components/native';

export const Login = () => {
  return (
    <StyledContainer>
      <StyledTitle>Login</StyledTitle>
    </StyledContainer>
  );
};

const StyledContainer = styled.SafeAreaView``;

const StyledTitle = styled.Text`
  font-size: ${({theme}) => theme.sizing.LARGE};
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
`;
