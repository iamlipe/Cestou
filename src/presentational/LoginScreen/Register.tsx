import React, {useEffect, useState} from 'react';
import styled from 'styled-components/native';
import * as Yup from 'yup';
import {SubmitHandler, useForm} from 'react-hook-form';
import {yupResolver} from '@hookform/resolvers/yup';
import {REGISTER, RegisterForm} from '@/store/slices/userSlice';
import {useReduxSelector} from '@/hooks/useReduxSelector';
import {useReduxDispatch} from '@/hooks/useReduxDispatch';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {AuthStackParamList} from '@/routes/stacks/AuthStack';
import {useTranslation} from 'react-i18next';
import {Keyboard} from 'react-native';

import RadioForm from '@/components/RadioForm';
import InputForm from '@/components/InputForm';
import CheckboxForm from '@/components/CheckboxForm';
import Button from '@/components/Button';

type NavProps = NativeStackNavigationProp<
  AuthStackParamList,
  'ConfirmRegister'
>;

export const Register = () => {
  const [canGoNext, setCanGoNext] = useState(false);
  const [registerData, setRegisterData] = useState({
    phoneOrEmail: '',
    password: '',
    userType: '',
  });
  const {goBack, navigate} = useNavigation<NavProps>();
  const {isLoading} = useReduxSelector(state => state.user);
  const {t} = useTranslation();
  const dispatch = useReduxDispatch();

  const schema = Yup.object().shape({
    userType: Yup.string().required(t('error.required')),
    name: Yup.string().required(t('error.required')),
    email: Yup.string()
      .min(10, t('error.validEmail'))
      .required(t('error.required')),
    password: Yup.string()
      .min(6, t('error.validPassword'))
      .required(t('error.required')),
    terms: Yup.array().min(1, t('error.terms')).required(),
  });

  const {
    control,
    handleSubmit,
    formState: {errors, isSubmitted},
  } = useForm<RegisterForm>({
    resolver: yupResolver(schema),
  });

  const onSubmit: SubmitHandler<RegisterForm> = async data => {
    const firstName = data.name.split(' ')[0];
    const lastName = data.name.split(' ')[1];
    const userType =
      data.userType === t('option.iConsumer') ? 'consumer' : 'producer';

    dispatch(
      REGISTER({
        firstName,
        lastName,
        userType,
        email: data.email,
        password: data.password,
      }),
    );

    Keyboard.dismiss();

    setRegisterData({
      phoneOrEmail: data.email,
      password: data.password,
      userType,
    });

    setCanGoNext(true);
  };

  useEffect(() => {
    if (!isLoading && canGoNext) {
      navigate('ConfirmRegister', registerData);
      setCanGoNext(false);
    }
  }, [isLoading, canGoNext, navigate, registerData]);

  return (
    <StyledContainer>
      <StyledRowTitle>
        <StyledTitle>{t('text.screenRegister.createAnAccount')}</StyledTitle>
        <StyledSubtitle>
          {t('text.screenRegister.enterYourDetails')}
        </StyledSubtitle>
      </StyledRowTitle>
      <StyledContainerForm>
        <RadioForm
          name="userType"
          control={control}
          options={[t('option.iProducer'), t('option.iConsumer')]}
        />

        <InputForm
          name="name"
          label={t('label.name')}
          error={isSubmitted ? errors.name?.message : ''}
          control={control}
        />
        <InputForm
          name="email"
          label={t('label.email')}
          error={isSubmitted ? errors.email?.message : ''}
          control={control}
        />
        <InputForm
          name="password"
          label={t('label.password')}
          error={isSubmitted ? errors.password?.message : ''}
          secureTextEntry
          control={control}
        />
        <StyledTextWarningPassowrd>
          {t('error.validPassword')}
        </StyledTextWarningPassowrd>

        <CheckboxForm
          name="terms"
          control={control}
          options={[t('option.terms')]}
        />
      </StyledContainerForm>

      <Button
        title={t('button.register')}
        loading={isLoading}
        onPress={handleSubmit(onSubmit)}
      />

      <StyledRowLogin>
        <StyledText>
          {t('text.screenRegister.alreadyHaveAnAccount?')}
        </StyledText>
        <StyledLink
          buttonColor="text_only"
          textColor="primary"
          title={t('button.signIn')}
          noMargin
          onPress={() => goBack()}
        />
      </StyledRowLogin>
    </StyledContainer>
  );
};

const StyledContainer = styled.ScrollView`
  min-height: 100%;
  background-color: ${({theme}) => theme.colors.BACKGROUND};
`;

const StyledContainerForm = styled.View`
  width: 90%;
  align-self: center;
  margin-bottom: 24px;
`;

const StyledRowTitle = styled(StyledContainerForm)`
  margin: 48px 0 0 0;
`;

const StyledTitle = styled.Text`
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.LARGEST};
  color: ${({theme}) => theme.colors.GRAY_900};
  margin-bottom: 12px;
`;

const StyledSubtitle = styled(StyledTitle)`
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  margin-bottom: 24px;
`;

const StyledRowLogin = styled.View`
  flex-direction: row;
  align-self: center;
  margin: 24px 0 48px;
`;

const StyledText = styled.Text`
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
`;

const StyledTextWarningPassowrd = styled(StyledText)`
  color: ${({theme}) => theme.colors.GRAY_900};
  margin: -16px 0 24px 0;
`;

const StyledLink = styled(Button)`
  width: auto;
  height: auto;
  margin: 0 5px;
`;
