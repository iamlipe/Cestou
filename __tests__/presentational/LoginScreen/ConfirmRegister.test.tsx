import React from 'react';
import {
  matchSnapshotWithProvider,
  renderWithThemeProvider,
} from '@/config/functions/Testing';

import {ConfirmRegister} from '@/presentational/LoginScreen/ConfirmRegister';

describe('ConfirmRegister', () => {
  test('should render screen correctly', () => {
    renderWithThemeProvider(<ConfirmRegister />);
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<ConfirmRegister />);
  });

  test('should render image welcome producer when loged with producer account', () => {
    const {getByTestId} = renderWithThemeProvider(<ConfirmRegister />, {
      user: {auth: {userType: 'producer'}},
    });

    const welcome = getByTestId('welcome-producer');

    expect(welcome).toBeTruthy();
  });

  test('should render image welcome consumer when loged with consumer account', () => {
    const {getByTestId} = renderWithThemeProvider(<ConfirmRegister />, {
      user: {auth: {userType: 'consumer'}},
    });

    const welcome = getByTestId('welcome-consumer');

    expect(welcome).toBeTruthy();
  });
});
