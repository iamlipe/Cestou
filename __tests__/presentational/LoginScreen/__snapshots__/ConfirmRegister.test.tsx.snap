// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmRegister should matches snapshot 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "alignItems": "center",
    }
  }
  showsVerticalScrollIndicator={false}
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "minHeight": "100%",
      },
    ]
  }
>
  <View>
    <SvgMock
      style={
        Array [
          Object {
            "marginTop": 40,
          },
        ]
      }
      testID="welcome-consumer"
    />
    <Text
      style={
        Array [
          Object {
            "color": "#00843F",
            "fontFamily": "SourceSansPro-SemiBold",
            "fontSize": 43,
            "marginTop": 40,
          },
        ]
      }
    >
      Cadastro realizado com sucesso!
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#1A1A1A",
            "fontFamily": "SourceSansPro-Regular",
            "fontSize": 27,
            "marginTop": 40,
            "textAlign": "center",
            "width": "90%",
          },
        ]
      }
    >
      Valide seu cadastro e aproveite toda a experiência de Cestar e ajudar no combate a fome.
    </Text>
    <View
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#00843F",
          "borderRadius": 4,
          "flexDirection": "row",
          "height": 49,
          "justifyContent": "center",
          "marginBottom": 10,
          "marginTop": 40,
          "opacity": 1,
          "width": 200,
        }
      }
      testID="confirm-button"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFFFFF",
              "fontFamily": "SourceSansPro-Regular",
              "fontSize": 27,
            },
          ]
        }
        textColor="white"
      >
        Começar
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
